### PR TITLE
feat(typing): add possibly undefined check to mypy

### PR DESCRIPTION
**Pull request by JoshFerge**
Unfortunately, in our codebase we sometimes encounter production errors, where a variable hasn't been bound. Such as https://sentry.sentry.io/issues/5000099518/.

We previously tried to use pyright for this, but running two type checkers isn't the greatest, so we removed it.
https://github.com/getsentry/sentry/pull/51350

Since mypy 1.0, there has been the optional error code `possibly-undefined`, which can detect these problems. I tried it out on our codebase, with around ~60 current issues in typed files. 

They mostly seem legitimate, and I confirmed this setting would have caught the above bug.

This PR enables it for all currently typed files, with an accompanying list of files currently typed that have possibly unbound variables. We can fix easier ones / ignore ones that would require a bigger refactor in follow up PRs.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -639,3 +648,51 @@ module = [
 ]
 disallow_untyped_defs = true
 # end: stronger typing
+
+# beginning: modules with possibly undefined variables
+[[tool.mypy.overrides]]
+module = [
+    "sentry.api.helpers.span_analysis",
```
Comment (bruno-garcia) [inline pyproject.toml:655]: Does this mean we need to add new modules manually?
Is there a 'blocklist' approach instead so new stuff gets here automagically?

Comment (JoshFerge) [reply pyproject.toml:655]: this is the blocklist approach -- these modules are being ignored at the moment as they have possibly unbound variable issues.

File: pyproject.toml [hunk 2]
```diff
@@ -64,6 +68,10 @@ warn_unused_configs = true
 warn_unused_ignores = true
 warn_redundant_casts = true
 
+# check for possibly unbound variables
+
```
Comment (asottile-sentry) [inline pyproject.toml:72]: this is a useless comment

File: pyproject.toml [hunk 3]
```diff
@@ -52,7 +52,11 @@ looponfailroots = ["src", "tests"]
 [tool.mypy]
 python_version = "3.11"
 mypy_path = ["fixtures/stubs-for-mypy"]
-plugins = ["pydantic.mypy", "mypy_django_plugin.main", "tools.mypy_helpers.plugin"]
+plugins = [
+    "pydantic.mypy",
+    "mypy_django_plugin.main",
+    "tools.mypy_helpers.plugin",
+]
```
Comment (asottile-sentry) [inline pyproject.toml:59]: unrelated to your change and not a helpful reformat imo

File: pyproject.toml [hunk 4]
```diff
@@ -639,3 +648,51 @@ module = [
 ]
 disallow_untyped_defs = true
 # end: stronger typing
+
+# beginning: modules with possibly undefined variables
+[[tool.mypy.overrides]]
+module = [
```
Comment (asottile-sentry) [inline pyproject.toml:654]: unless you're going to make an autofixer for this we shouldn't do this.  let's fix the issues first and then enable the error code

**Review by asottile-sentry — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 3 above)
Comment (JoshFerge) [reply pyproject.toml:59]: would be nice to have a TOML formatter in pre-commit as my default toml formatter always makes this change. i've turned off my TOML formatter for future PRs here

**Comment by JoshFerge** [comment]
Converting to draft, while I fix the errors already in the codebase. will put up separate PRs for them.

File: pyproject.toml (hunk 4 above)
Comment (asottile-sentry) [reply pyproject.toml:654]: yeah I mean more something which whittles this allowlist down automatically (how the rest of the allowlist works) -- but I think we should fix the problems first anyway

Comment (JoshFerge) [reply pyproject.toml:654]: ah i see what you're syaing

**Review by asottile-sentry — CHANGES_REQUESTED** [review]
I'd rather we fix these instead of another blocklist when we're not even done with the first one

**Comment by untitaker** [comment]
instead of adding a blocklist, would it make sense to add `type: ignore` to each line exhibiting the error? it's basically the same thing except the blocklist is line-wise instead of file-wise

**Comment by asottile-sentry** [comment]
> instead of adding a blocklist, would it make sense to add `type: ignore` to each line exhibiting the error? it's basically the same thing except the blocklist is line-wise instead of file-wise

my fear is that they'll never get fixed and are lurking bugs waiting to cause incidents -- so we should feel empowered to fix them now

**Comment by JoshFerge** [comment]
I can definitely fix the easier ones, and get the hard ones prioritized -- this effort will still end up taking a bit though, if i had to guess. I get that its annoying to add another blocklist, but in the end I think this will help prevent incidents pretty immediately, and i don't really see a downside to it.

we can help validate code immediately and get fixing the existing ones prioritized.

**Pull request closed without merging**